### PR TITLE
add event triggers

### PR DESCRIPTION
**Pull request by tirumaraiselvan**

### CODE REVIEW

File: server/src-rsr/initialise.sql [hunk 1]
```diff
@@ -179,3 +179,51 @@ LANGUAGE plpgsql AS $$
       END LOOP;
     END;
 $$;
+
+CREATE TABLE hdb_catalog.event_triggers
+(
+  id SERIAL UNIQUE,
+  name TEXT PRIMARY KEY,
```
Comment (0x777) [inline server/src-rsr/initialise.sql]: id should be primary key, name should have unique

**Comment by shahidhk** [comment]
@praveenweb Creating a new trigger redirects to 404 on success

**Comment by coco98** [comment]
@praveenweb Also this bug where I don't know what the error is. It's in the name part, but was hard for me to figure out.

screen shot 2018-08-30 at 5 18 42 pm

**Comment by shahidhk** [comment]
@praveenweb Couple of other feedback:

1. 
image

`s/Select operations/Operations`

2. 
Validate the webhook url to be a url (I'm guessing there is a input type url?)

**Comment by ecthiender** [comment]
@tirumaraiselvan also this has to be a catalog version update. Otherwise ppl can't upgrade

**Comment by shahidhk** [comment]
I created a trigger, added new column to table, and tried deleting a column and it failed as expected, citing dependent trigger.

Then I tried to delete the trigger, but this happened:
```json
{
    "statement": "DROP FUNCTION IF EXISTS hdb_views.notify_skor_profile_change_INSERT CASCADE",
    "prepared": false,
    "error": {
        "exec_status": "FatalError",
        "hint": null,
        "message": "syntax error at or near \"CASCADE\"",
        "status_code": "42601",
        "description": null
    },
    "arguments": []
}
```

**Comment by shahidhk** [comment]
Also, I notice that while inserting, I am receiving an update event too!

**Comment by shahidhk** [comment]
@praveenweb The up and down migrations created for a trigger is the same:

```yaml
- args:
    insert:
      columns:
      - id
      - name
      - lng
      - lat
      - address
    name: new_profile
    table:
      name: profile
      schema: public
    webhook: https://us-central1-hasura-test.cloudfunctions.net/trigger
  type: subscribe_table
```

File: server/src-rsr/initialise.sql (hunk 1 above)
Comment (tirumaraiselvan) [reply server/src-rsr/initialise.sql]: I am not using id anywhere, I can just remove it.

File: server/src-lib/Hasura/Events/Lib.hs [hunk 2]
```diff
@@ -0,0 +1,277 @@
...
+  , eeCtxPollingIntervalSec :: Int
+  }
+
+defMaxEventThreads :: Int
+defMaxEventThreads = 100
+
+defPollingIntervalSec :: Int
+defPollingIntervalSec = 5
+
+initEventEngineCtx :: Int -> Int -> STM (TVar EventEngineCtx)
```
Comment (0x777) [inline server/src-lib/Hasura/Events/Lib.hs]: why is eventenginectx wrapped in tvar?

File: server/src-lib/Hasura/RQL/DDL/Metadata.hs [hunk 3]
```diff
@@ -214,6 +221,11 @@ applyQP2 (ReplaceMetadata tables templates) = do
       withPathK "delete_permissions" $ processPerms tabInfo $
         table ^. tmDeletePermissions
 
+    indexedForM_ tables $ \table -> do
```
Comment (0x777) [inline server/src-lib/Hasura/RQL/DDL/Metadata.hs]: what is this code doing under `withPathK "queryTemplates"`

File: server/src-lib/Hasura/Events/Lib.hs [hunk 4]
```diff
@@ -0,0 +1,277 @@
...
+consumeEvents
+  :: HLogger -> HTTPSessionMgr -> Q.PGPool -> CacheRef -> TVar EventEngineCtx -> IO ()
+consumeEvents logger httpSess pool cacheRef eectx  = forever $ do
+  event <- atomically $ do
+    EventEngineCtx q _ _ _ <- readTVar eectx
+    TQ.readTQueue q
+  putStrLn "got event"
+  async $ runReaderT  (processEvent pool event) (logger, httpSess, cacheRef, eectx)
+
+processEvent
```
Comment (0x777) [inline server/src-lib/Hasura/Events/Lib.hs]: split this function

Comment (tirumaraiselvan) [reply server/src-lib/Hasura/Events/Lib.hs]: done

File: server/src-lib/Hasura/RQL/DDL/Metadata.hs (hunk 3 above)
Comment (tirumaraiselvan) [reply server/src-lib/Hasura/RQL/DDL/Metadata.hs]: expand the PR bruh

File: server/src-rsr/trigger.sql [hunk 5]
```diff
@@ -0,0 +1,31 @@
+CREATE OR REPLACE function hdb_views.notify_skor_{{NAME}}_{{OPERATION}}() RETURNS trigger
+   LANGUAGE plpgsql
+   AS $$
+   DECLARE
+   payload json;
+   data json;
+   id text;
+   BEGIN
+     raise notice 'pushing to event log';
```
Comment (0x777) [inline server/src-rsr/trigger.sql]: this shouldn't be present in prod mode

File: server/stack.yaml [hunk 6]
```diff
@@ -19,6 +19,8 @@ extra-deps:
   commit: e61bc37794b4d9e281bad44b2d7c8d35f2dbc770
 - git: https://github.com/hasura/graphql-parser-hs.git
   commit: eae59812ec537b3756c3ddb5f59a7cc59508869b
+- git: https://github.com/hasura/ginger.git
```
Comment (0x777) [inline server/stack.yaml]: this is fixed upstream. https://github.com/tdammers/ginger

File: server/src-exec/Main.hs [hunk 7]
```diff
@@ -176,6 +194,11 @@ main =  do
       currentTime <- getCurrentTime
       res <- runTx ci $ migrateCatalog currentTime
       either ((>> exitFailure) . printJSON) putStrLn res
+    prepareEvents ci = do
+      putStrLn "event_triggers: preparing data"
```
Comment (0x777) [inline server/src-exec/Main.hs]: Is this needed?

Comment (tirumaraiselvan) [reply server/src-exec/Main.hs]: why not? it's a startup step right?

File: server/src-exec/Main.hs [hunk 8]
```diff
@@ -176,6 +193,24 @@ main =  do
...
+      res <- runTx ci unlockAllEvents
+      either ((>> exitFailure) . printJSON) return res
+    getMaxEventThreads = do
+      mEnv <- lookupEnv "HASURA_GRAPHQL_EVENTS_HTTP_POOL_SIZE"
```
Comment (0x777) [inline server/src-exec/Main.hs]: can be abstracted out in a function:
```haskell
getFromEnv :: (Read a) => String -> IO a
```

File: server/src-lib/Hasura/Events/HTTP.hs [hunk 9]
```diff
@@ -0,0 +1,362 @@
...
+  = ExtraContext
+  { elEventCreatedAt :: Time.UTCTime
+  , elEventId        :: TriggerId
+  } deriving (Show, Eq)
+
+$(J.deriveJSON (J.aesonDrop 2 J.snakeCase){J.omitNothingFields=True} ''ExtraContext)
+
+data HTTPSessionMgr
+  = HTTPSessionMgr
+  { _hsmSession         :: !WS.Session
```
Comment (0x777) [inline server/src-lib/Hasura/Events/HTTP.hs]: Get rid of this type. Just use WS.Session. Let's not support insecure connections to webhooks.

File: server/src-lib/Hasura/Events/HTTP.hs [hunk 10]
```diff
@@ -0,0 +1,362 @@
...
+  res <- R.retrying retryPol' retryFn' $ httpWithLogging opts True http exLog
+
+  -- process the result
+  either throwError return res
+
+  where
+    retryPol'  = R.RetryPolicyM $ liftIO . R.getRetryPolicyM (_hRetryPolicy http)
+    retryFn' _ = return . _hRetryFn http
+
+runInsecureHTTP
```
Comment (0x777) [inline server/src-lib/Hasura/Events/HTTP.hs]: This function is not needed

File: server/src-lib/Hasura/Events/HTTP.hs [hunk 11]
```diff
@@ -0,0 +1,362 @@
...
+  -- process the result
+  either throwError return res
+  where
+    retryPol'  = R.RetryPolicyM $ liftIO . R.getRetryPolicyM (_hRetryPolicy http)
+    retryFn' _ = return . _hRetryFn http
+
+
+httpWithLogging
+  :: (MonadReader r m, MonadIO m, Has HTTPSessionMgr r, Has HLogger r)
+  => W.Options -> Bool -> HTTP a -> Maybe ExtraContext -> R.RetryStatus -> m (Either HTTPErr a)
```
Comment (0x777) [inline server/src-lib/Hasura/Events/HTTP.hs]: remove the `isSecure` argument

File: server/src-lib/Hasura/Events/HTTP.hs [hunk 12]
```diff
@@ -0,0 +1,362 @@
...
+  return $ either (Left . HClient) bodyParser res
+
+  where
+    -- set wreq options to ignore status code exceptions
+    ignoreStatusCodeExceptions _ _ = return ()
+    finalOpts = optsMod opts
+                & W.checkResponse .~ Just ignoreStatusCodeExceptions
+
+    -- the actual function which makes the relevant Wreq calls
+    finallyRunHTTPPlz sessMgr =
```
Comment (0x777) [inline server/src-lib/Hasura/Events/HTTP.hs]: `Plz`?

File: server/src-rsr/trigger.sql [hunk 13]
```diff
@@ -0,0 +1,26 @@
+CREATE OR REPLACE function hdb_views.notify_hasura_{{NAME}}_{{OPERATION}}() RETURNS trigger
```
Comment (0x777) [inline server/src-rsr/trigger.sql]: `.sql.j2` extension

File: server/src-rsr/initialise.sql [hunk 14]
```diff
@@ -179,3 +179,48 @@ LANGUAGE plpgsql AS $$
       END LOOP;
     END;
 $$;
+
+-- required for generating uuid
+CREATE EXTENSION IF NOT EXISTS pgcrypto;
+
+CREATE TABLE hdb_catalog.event_triggers
```
Comment (0x777) [inline server/src-rsr/initialise.sql]: follow the current convention for table names, i.e, we use singulars.

File: server/src-lib/Hasura/RQL/DDL/Utils.hs [hunk 15]
```diff
@@ -2,14 +2,38 @@
...
+   \ END $$; "
+
+
+clearHdbViewsFunc :: BB.Builder
+clearHdbViewsFunc =
+  "DO $$ DECLARE \
+  \ _sql text; \
+  \ BEGIN \
+  \ SELECT INTO _sql \
+  \        string_agg(format('DROP FUNCTION %s(%s) CASCADE;' \
```
Comment (0x777) [inline server/src-lib/Hasura/RQL/DDL/Utils.hs]: You'll have to use `quote_ident` for identifiers.

**Pull request merged**